lib/html_diff/differ.rb

```diff
@@ -0,0 +1,78 @@
+require 'nokogiri'
+require 'diffy'
+
+module HtmlDiff
+  class Differ
+    REPLACEMENTS = {
+      'https://www-origin.staging.publishing.service.gov.uk' => 'https://www.gov.uk',
+      'https://www-origin.publishing.service.gov.uk' => 'https://www.gov.uk',
+      'https://assets-origin.staging.publishing.service.gov.uk' => 'https://assets.digital.cabinet-office.gov.uk',
+      /https:\/\/assets\.digital\.cabinet-office\.gov\.uk\/specialist-frontend\/application-[0-9a-f]{32}\.js/ => 'https://assets.digital.cabinet-office.gov.uk/specialist-frontend/application-7463fa64f198b6568dc121dae41d44b1.js',
+    }
+
+    attr_reader :differing_pages
+
+    def initialize(config)
+      @config = config
+      @template = File.read "#{ROOT_DIR}/lib/html_diff/assets/html_diff_template.erb"
+      @diff_dir = "#{ROOT_DIR}/#{@config.html_diff.directory}"
+      reset_html_diffs_dir
+      @differing_pages = {}
+    end
+
+    def diff(base_path)
+      staging_html = get_normalized_html(staging_url(base_path))
+      production_html = get_normalized_html(production_url(base_path))
+      diffy = Diffy::Diff.new(production_html, staging_html, context: 3)
+      unless diffy.diff == ""
+        write_diff_page(base_path, diffy.to_s(:html))
+        @differing_pages[base_path] = html_diff_filename(base_path)
```

Comment:
But I don't need a setter. Getters and Setters are only necessary when the instance varibable need to be set/got from outside the class, otherwise we should use the instance variable name from within the class. In this case, I need to get the variable from outside the class, but never set it from outside the class. By specifying a getter, but no setter, we sending an important signal: This is a variable that can only be manipulated from within this class, although we can read it from outside the class.